server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -2179,7 +2179,7 @@ private boolean shouldPeriodicallyFlush(long flushThresholdSizeInBytes, long flu
     }
 
     @Override
-    public void flush(boolean force, boolean waitIfOngoing, ActionListener<FlushResult> listener) throws EngineException {
+    protected void flushHoldingLock(boolean force, boolean waitIfOngoing, ActionListener<FlushResult> listener) throws EngineException {
         ensureOpen();
```

Comment:
Can we assert that the current thread effectively holds a lock on the read or write lock?

Reply:
Strictly speaking yes we can, for now, but https://github.com/elastic/elasticsearch/pull/103123 will break the relationship between threads and locks anyway so I am inclined not to add that assertion here.